src/btree/bt_misc.c

```diff
@@ -139,3 +139,172 @@ __wt_page_type_string(u_int type) WT_GCC_FUNC_ATTRIBUTE((visibility("default")))
     }
     /* NOTREACHED */
 }
+
+/*
+ * __wt_page_is_root --
+ *     Returns true if the page is a root page.
+ */
+static WT_INLINE bool
+__wt_page_is_root(WT_PAGE *page)
+{
+    return (page->pg_intl_parent_ref == NULL || page->pg_intl_parent_ref->home == NULL);
+}
+
+/*
+ * __wt_page_npos --
+ *     Get the page's normalized position in the tree.
+ */
+double
+__wt_page_npos(WT_SESSION_IMPL *session, WT_REF *ref)
+{
+    WT_PAGE *page, *parent_page;
+    WT_REF *self_ref;
+    double npos;
+    uint32_t idx, entries;
+
+    npos = 0.5;
+    page = ref->home;
+    if (!page)
+        return (npos);
+
+    for (;;) {
+        WT_ASSERT(session, page != NULL);
+        WT_ASSERT(session, WT_PAGE_IS_INTERNAL(page));
+
+        self_ref = page->pg_intl_parent_ref; /* The field name is deceptive */
+        parent_page = self_ref->home;
+
+        idx = ref->pindex_hint;
+        // ?? should be used with WT_WITH_PAGE_INDEX and WT_INTL_INDEX_GET()?
+        entries = WT_INTL_INDEX_GET_SAFE(page)->entries;
+
+        if (idx < entries)
+            npos = (idx + npos) / entries;
+        /*
+         * If idx is incorrect, just leave the position as it is. We can scan all entries for the
+         * correct position, but it's not worth it.
+         */
+
+        if (!parent_page || page == parent_page || __wt_page_is_root(page))
```

Comment:
Couple of nits:
`!parent_page` -> `parent_page != NULL`
What cases will `page == parent_page` catch? If a page contains itself as a child this seems like a serious problem